tests/test_cli_compile.py

```diff
@@ -2702,9 +2706,225 @@ def test_all_extras_fail_with_extra(fake_dists, runner, make_module, fname, cont
     assert exp in out.stderr
 
 
+def _mock_resolver_cls(monkeypatch: MonkeyPatch) -> MagicMock:
+    obj = MagicMock()
+    obj.resolve = MagicMock(return_value=set())
+    obj.resolve_hashes = MagicMock(return_value=dict())
+    cls = MagicMock(return_value=obj)
+
+    monkeypatch.setattr("piptools.scripts.compile.BacktrackingResolver", cls)
+    monkeypatch.setattr("piptools.scripts.compile.LegacyResolver", cls)
+
+    return cls
+
+
+def _mock_build_project_metadata(monkeypatch: MonkeyPatch) -> MagicMock:
+    func = MagicMock(
+        return_value=ProjectMetadata(
+            extras=("e",),
+            requirements=(
+                install_req_from_line("rdep0"),
+                install_req_from_line("rdep1; extra=='e'"),
+            ),
+            build_requirements=(install_req_from_line("bdep0"),),
+        )
+    )
+
+    monkeypatch.setattr("piptools.scripts.compile.build_project_metadata", func)
+
+    return func
```

Comment:
How can `mocks` be avoided in integration tests? What problem does `mocks` solve in newly added tests?

Reply:
The problem the `mock`s solve is avoiding running functions that have no impact on the functionality being tested but take a long time. I can think of a few ways to avoid them

1. run the entire `cli` and accept the longer test duration
2. refactor `cli` to allow the functionality to be tested without invoking `cli` or using monkey patching
3. remove the tests and accept the lower test coverage
5. remove the functionality being tested

`test_all_build_distributions` essentially ensures the existence of and provides coverage for this branch

```python
elif all_build_deps:
    build_deps_for_distributions = ALL_BUILD_DISTRIBUTIONS
```

`test_only_build_distributions` essentially ensures the existence of this branch

```python
if not build_deps_only:
    constraints.extend(metadata.requirements)
    if all_extras:
        if extras:
            msg = "--extra has no effect when used with --all-extras"
            raise click.BadParameter(msg)
        extras = metadata.extras
```

Reply:
Thanks for the info! I'll try to get back with a solution shortly.